formats/json/jvmMain/src/kotlinx/serialization/json/JvmStreams.kt

```diff
@@ -0,0 +1,76 @@
+package kotlinx.serialization.json
+
+import kotlinx.serialization.*
+import kotlinx.serialization.json.internal.*
+import java.io.*
+import java.nio.charset.Charset
+
+/**
+ * Serializes the [value] with [serializer] into a [stream] using JSON format and given [charset].
+ *
+ * @throws [SerializationException] if the given value cannot be serialized to JSON.
+ * @throws [IOException] If an I/O error occurs and stream can't be written to.
+ */
+@ExperimentalSerializationApi
+public fun <T> Json.encodeToStream(
+    serializer: SerializationStrategy<T>,
+    value: T,
+    stream: OutputStream,
+    charset: Charset = Charsets.UTF_8
+) {
+    val result = JsonToWriterStringBuilder(stream, charset)
+    try {
+        val encoder = StreamingJsonEncoder(
+            result, this,
+            WriteMode.OBJ,
+            arrayOfNulls(WriteMode.values().size)
+        )
+        encoder.encodeSerializableValue(serializer, value)
+    } finally {
+        result.release()
+    }
+}
+
+/**
+ * Serializes given [value] to [stream] using [charset] and serializer retrieved from the reified type parameter.
+ *
+ * @throws [SerializationException] if the given value cannot be serialized to JSON.
+ * @throws [IOException] If an I/O error occurs and stream can't be written to.
+ */
+@ExperimentalSerializationApi
+public inline fun <reified T> Json.encodeToStream(
+    value: T,
+    stream: OutputStream,
+    charset: Charset = Charsets.UTF_8
```

Comment:
A charset is quite a serious commitment in our public API.
Let's stick with UTF-8 only and document it as well?

Reply:
I don't see many problems in it — Charset isn't going anywhere from `java.nio`, AFAIK. We still can implement charset-specific decoders if we need to.

Of course, we also can add overload with charset later, if there would be demand for it, but we'll end up with two almost identical overloads instead of a function with default parameter

Reply:
It will obstruct us from properly optimizing UTF-8 decoding because we will be forced to maintain the rest of the charsets (think about suffix tree), apart from that charsets are quite slow on their own.

Moreover, most of the modern applications are using UTF-8 anyway and it's unlikely we'll stumble across such demand. So I suggest avoiding introducing charsets into our API and rely on UTF-8 only, waiting for a proper demand.

Reply:
@qwwdfsad I agree on the character set thing. Especially in the context of taking a "writer". The api to handle the charsets is already there. The only counterpoint is the speed of parsing streams vs readers.

Reply:
Good to know, thanks!

Leonid is on the vacation right now, so the final decision is not yet made, but we've discussed this and other cases on our design meeting and seems that we have a consensus regarding encodings -- UTF-8 dominates all along, and everyone is mostly fine to be UTF-8 first. Even Java changes towards that: [JEP-400](https://openjdk.java.net/jeps/400). Moshi is also a good example of that.

So it's _likely_ that all our new API will be using UTF-8 as well and other encoding will be put aside and be supported on a best-effort basis